src/mds/MDLog.cc

```diff
@@ -651,13 +656,32 @@ void MDLog::trim(int m)
     ceph_assert(segments.size() >= pre_segments_size);
     max_expiring_segments = std::max<unsigned>(max_expiring_segments,segments.size() - pre_segments_size);
   }
-  
+
   map<uint64_t,LogSegment*>::iterator p = segments.begin();
+
+  auto trim_start = ceph::coarse_mono_clock::now();
+  std::optional<ceph::coarse_mono_time> trim_end;
+
+  auto log_trim_counter_start = log_trim_counter.get();
+  auto log_trim_threahold = g_conf().get_val<Option::size_t>("mds_log_trim_threshold");
+
   while (p != segments.end()) {
-    if (stop < ceph_clock_now())
-      break;
+    // throttle - break out of trimmming if we've hit the threahold
+    if (log_trim_counter_start + new_expiring_segments >= log_trim_threahold) {
+      auto time_spent = std::chrono::duration<double>::zero();
+      if (trim_end) {
+	time_spent = std::chrono::duration<double>(*trim_end - trim_start);
+      }
+      dout(10) << __func__ << ": breaking out of trim loop - trimmed "
+	       << new_expiring_segments << " segment(s) in " << time_spent.count()
+	       << "s" << dendl;
+    }
 
     unsigned num_remaining_segments = (segments.size() - expired_segments.size() - expiring_segments.size());
+    dout(10) << __func__ << ": new_expiring_segments=" << new_expiring_segments
+	     << ", num_remaining_segments=" << num_remaining_segments
+	     << ", max_segments=" << max_segments << dendl;
+
     if ((num_remaining_segments <= max_segments) &&
 	(max_events < 0 || num_events - expiring_events - expired_events <= max_events))
       break;
```

Comment:
I think we should also remove this:

`if (new_expiring_segments * 2 > num_remaining_segments)`

I don't see a reason why the MDS shouldn't just trim segments up to the current segment. i.e. if all but the latest segment is expired, that's perfectly good IMO.

Reply:
This may disrupt tests however because some test may not expect RADOS objects for .e.g directories to be created until a `journal flush`. It's worth a QA run to see though!

Reply:
Right. Those heuristics drive me crazy every time I see this code. Perfect time to strike if off...

Reply:
And, the following too is a candidate for removal:

```
    if (max_expiring_segments > 0 &&
        expiring_segments.size() >= max_expiring_segments)
      break;
```